### PR TITLE
Implement DRClusterConfig reconciler to create required ClusterClaims

### PR DESCRIPTION
**Pull request by ShyamsundarR**
Part of https://github.com/RamenDR/ramen/pull/1403

This implements creation of ClusterClaims for the various classes detected on the ManagedCluster that is also marked via a DRClusterConfig as a DR peer.

Future considerations:

- Add a basic status (say status.Message and observedGeneration) to DRClusterConfig reflecting success or failure of reconciliation
- Add required CEL immutability rules to DRClusterConfig, as certain fields should not be changed once created (e.g `clusterID`)
- DRClusterConfig can check s3 accessibility and report issues regarding the same, closing the gap for S3 validation on the ManagedCluster
- Add any required `clusterID` validation

### CODE REVIEW

File: internal/controller/drclusters.go [hunk 1]
```diff
@@ -245,6 +245,14 @@ func drClusterUndeploy(
 		return fmt.Errorf("drcluster '%v' referenced in one or more existing drPolicy resources", drcluster.Name)
 	}
 
+	if err := mwu.DeleteManifestWork(mwu.BuildManifestWorkName(util.MWTypeDRCConfig), drcluster.GetName()); err != nil {
+		return err
+	}
+
+	if _, err := mwu.FindManifestWork(mwu.BuildManifestWorkName(util.MWTypeDRCConfig), drcluster.GetName()); err == nil {
```
Comment (ELENAGER) [inline internal/controller/drclusters.go:252]: @ShyamsundarR it's the only place we are checking manifest after deletion. If there is no error in line [248](https://github.com/RamenDR/ramen/pull/1485/files#diff-0fe31ec47a23104282e03d05496277a6352b98ffae3fcab8a4e8b7d24c9f534eR248:~:text=if%20err%20%3A%3D%20mwu.DeleteManifestWork(mwu.BuildManifestWorkName(util.MWTypeDRCConfig)%2C%20drcluster.GetName())%3B%20err%20!%3D%20nil%20%7B), the manifest will be deleted, so why we need an additional check?

Comment (ShyamsundarR) [reply internal/controller/drclusters.go:252]: Good question, I want to ensure that the DRClusterConfig resource at the managed cluster is deleted before we proceed to delete the MW for the ramen components at the managed cluster. If we reach that before the resource is actually deleted, then the DRClusterConfig reconciler may not be existing to finalize the resource locally.

Hence we delete and wait till it is actually deleted, then we proceed with deletion of the ramen operators on the managed cluster.

File: internal/controller/util/mw_util.go [hunk 2]
```diff
@@ -298,6 +364,28 @@ func Namespace(name string) *corev1.Namespace {
...
 }
 
+func ExtractResourceFromManifestWork(
+	mw *ocmworkv1.ManifestWork,
+	object client.Object,
+	gvk schema.GroupVersionKind,
+) error {
+	rawObject, err := GetRawExtension(mw.Spec.Workload.Manifests, gvk)
+	if err != nil {
+		return fmt.Errorf("failed fetching MaintenanceMode from manifest %w", err)
```
Comment (ELENAGER) [inline internal/controller/util/mw_util.go:374]: failed fetching resource from manifest? Not MaintenanceMode

File: internal/controller/util/mw_util.go [hunk 3]
```diff
@@ -298,6 +364,28 @@ func Namespace(name string) *corev1.Namespace {
...
+	}
+
+	if rawObject == nil {
+		return nil
+	}
+
+	err = json.Unmarshal(rawObject.Raw, object)
+	if err != nil {
+		return fmt.Errorf("failed unmarshaling MaintenanceMode from manifest %w", err)
```
Comment (ELENAGER) [inline internal/controller/util/mw_util.go:383]: Same as previous

File: internal/controller/drcluster_controller.go [hunk 4]
```diff
@@ -428,13 +423,27 @@ func (r DRClusterReconciler) processCreateOrUpdate(u *drclusterInstance) (ctrl.R
...
+
+		u.log.Info("Error during processing maintenance modes", "error", err)
+	}
+
 	if err := u.statusUpdate(); err != nil {
 		u.log.Info("failed to update status", "failure", err)
 	}
 
-	return ctrl.Result{Requeue: requeue || u.requeue}, reconcileError
+	return ctrl.Result{Requeue: requeue || u.requeue}, nil
```
Comment (ELENAGER) [inline internal/controller/drcluster_controller.go:446]: If there is an error during fencing - we will not return error?

File: internal/controller/drcluster_controller.go [hunk 5]
```diff
@@ -615,6 +624,80 @@ func (u *drclusterInstance) finalizerRemove() error {
...
+	drpolicies, err := util.GetAllDRPolicies(u.ctx, u.reconciler.APIReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that schedules are not duplicated by, storing them in "added" to avoid adding a duplicate schedule from
+	// another DRPolicy
+	added := map[string]bool{}
+
+	for idx := range drpolicies.Items {
```
Comment (ELENAGER) [inline internal/controller/drcluster_controller.go:678]: Is it possible, that policy is marked for deletion? Do we need to handle it anyway?

File: internal/controller/util/mw_util.go (hunk 2 above)
Comment (ShyamsundarR) [reply internal/controller/util/mw_util.go]: Fixed

File: internal/controller/util/mw_util.go (hunk 3 above)
Comment (ShyamsundarR) [reply internal/controller/util/mw_util.go]: Fixed

File: internal/controller/drcluster_controller.go (hunk 4 above)
Comment (ShyamsundarR) [reply internal/controller/drcluster_controller.go:446]: If there is an error in fencing we will note that in the status and log it, and we set requeue to true (for future reconciles). Returning an error is not required at this point.

File: internal/controller/drcluster_controller.go (hunk 5 above)
Comment (ShyamsundarR) [reply internal/controller/drcluster_controller.go]: Interestingly there is a test case for this that passes, though based on the comment I think it is possible that we may retain a stale schedule till a future DRCluster reconcile (due to any reason) is initiated. Will fix this by skipping deleted DRPolicies in the loop.

The parent PR that this is a part of is merged, hence will address it as a separate PR.

**Review by BenamarMk — APPROVED** [review]
LGTM

**Pull request merged**